atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/BatchSizeIncreasingIterator.java

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright 2015 Palantir Technologies
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.transaction.impl;
+
+import java.util.List;
+
+import javax.annotation.Nullable;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterators;
+import com.palantir.atlasdb.AtlasDbPerformanceConstants;
+import com.palantir.common.base.ClosableIterator;
+import com.palantir.common.base.ClosableIterators;
+import com.palantir.util.AssertUtils;
+
+interface BatchProvider<T> {
+    ClosableIterator<T> getBatch(int batchSize, @Nullable byte[] lastToken);
+    boolean hasNext(byte[] lastToken);
+    byte[] getLastToken(List<T> batch);
+}
+
+class BatchSizeIncreasingIterator<T> {
+    final int originalBatchSize;
+
+    final BatchProvider<T> batchProvider;
+    ClosableIterator<T> currentResults = null;
+    byte[] lastToken;
+
+    long numReturned = 0;
+    long numNotDeleted = 0;
+    int lastBatchSize;
+
+    public BatchSizeIncreasingIterator(BatchProvider<T> batchProvider,
+                                       int originalBatchSize,
+                                       @Nullable ClosableIterator<T> currentResults) {
+        Preconditions.checkArgument(originalBatchSize > 0);
+        this.batchProvider = batchProvider;
+        this.originalBatchSize = originalBatchSize;
+    }
+
+    public void markNumResultsNotDeleted(int resultsInBatch) {
+        numNotDeleted += resultsInBatch;
+        AssertUtils.assertAndLog(numNotDeleted <= numReturned, "NotDeleted is bigger than the number of results we returned.");
+    }
+
+    int getBestBatchSize() {
+        if (numReturned == 0) {
+            return originalBatchSize;
+        }
+        final long batchSize;
+        if (numNotDeleted == 0) {
+            // If everything we've seen has been deleted, we should be aggressive about getting more rows.
+            batchSize = numReturned * 4;
```

Comment:
nit: The choice of 4 here is a little arbitrary (which is fine), but I guess it feels a little weird since this can be less of an increase than numNotDelete != 0.  Perhaps the else should be the min of what it currently is and `numReturned * 4` (the batch rate of change is then capped).